### PR TITLE
feat:  top level withdrawal methods [OTE-389] [3/n]

### PR DESCRIPTION
**Pull request by yogurtandjam**
Followup to https://github.com/dydxprotocol/v4-abacus/pull/443

Enables exchange and non-cctp deposits + withdrawals.

This PR:
- adds the concept of a `skipDenom` as the skip denom for tokens on their native chains differs from what is expected in our frontend SDK's. We need to keep a record of both so we can successfully submit requests to the skip endpoint and commit the tx via our frontend SDK's.
- Adds osmosis and neutron addresses to the skip withdrawal/deposit non-cctp api requests to enable more routes to keep us as close to parity with existing squid integration
- adds calculation to skipRouteProcessor that iterates through the `operations` property on the skip `msgs_direct` response payload to get the bridge fees. it turns out the `estimated_fees` key is not comprehensive
- we've set `allow_unsafe_tx` to `true` to keep user experience at parity w/ squid. that being said this is possibly dangerous and as part of [cleanup](https://linear.app/dydx/issue/OTE-350/%5Babacus%5D-cleanup) and QA we may choose to remove this
- removes some comments about possibly deduplicating code. we can worry about that in the future. right now the differences enough in each endpoint calling method are significant enough where the abstraction opportunity isn't immediately obvious.

QA'd Flows:
- Cctp deposit
- Cctp withdrawal
- Cctp auto sweep
- Non cctp deposit
- Non cctp deposit
- Non cctp withdrawal
- Non cctp auto sweep
- Exchange deposit
- Exchange withdrawal

### CODE REVIEW

File: src/commonMain/kotlin/exchange.dydx.abacus/processor/router/skip/SkipProcessor.kt [hunk 1]
```diff
@@ -174,7 +179,31 @@ internal class SkipProcessor(
     override fun filteredTokens(chainId: String?): List<Any>? {
         val chainIdToUse = chainId ?: defaultChainId()
         val assetsMapForChainId = parser.asNativeMap(this.skipTokens?.get(chainIdToUse))
-        return parser.asNativeList(assetsMapForChainId?.get("assets"))
+        val assetsForChainId = parser.asNativeList(assetsMapForChainId?.get("assets"))
+//      coinbase exchange chainId is noble-1. we only allow usdc withdrawals from it
+        if (chainId === "noble-1") {
```
Comment (yogurtandjam) [inline src/commonMain/kotlin/exchange.dydx.abacus/processor/router/skip/SkipProcessor.kt:184]: fixed in folllowup PR

File: src/commonMain/kotlin/exchange.dydx.abacus/processor/router/skip/SkipRouteProcessor.kt [hunk 2]
```diff
@@ -32,26 +38,47 @@ internal class SkipRouteProcessor(internal val parser: ParserProtocol) {
...
+        val operations = parser.asList(parser.value(payload, "route.operations")) ?: return 0.0
+        var total = 0.0
+        operations.forEach { operation ->
+            val fee = getFeeFromOperation(parser.asNativeMap(operation))
+            total += fee
+        }
+        return total
+    }
+
+    private fun getFeeFromOperation(operationPayload: Map<String, Any>?): Double {
```
Comment (yogurtandjam) [inline src/commonMain/kotlin/exchange.dydx.abacus/processor/router/skip/SkipRouteProcessor.kt:51]: [fix] make a comment describing this

File: src/commonMain/kotlin/exchange.dydx.abacus/state/model/TradingStateMachine+TransferInput.kt [hunk 3]
```diff
@@ -245,7 +246,9 @@ private fun TradingStateMachine.updateTransferToChainType(transfer: MutableMap<S
 
 private fun TradingStateMachine.updateTransferExchangeType(transfer: MutableMap<String, Any>, exchange: String) {
     val exchangeDestinationChainId = routerProcessor.exchangeDestinationChainId
+    Logger.e({ "exchangedestinationchainid:$exchangeDestinationChainId" })
```
Comment (yogurtandjam) [inline src/commonMain/kotlin/exchange.dydx.abacus/state/model/TradingStateMachine+TransferInput.kt:249]: [fix] remove this

File: src/commonMain/kotlin/exchange.dydx.abacus/state/model/TradingStateMachine+TransferInput.kt [hunk 4]
```diff
@@ -245,7 +246,9 @@ private fun TradingStateMachine.updateTransferToChainType(transfer: MutableMap<S
...
     val tokenOptions = routerProcessor.tokenOptions(exchangeDestinationChainId)
+    Logger.e({ "tokenOptions:$tokenOptions" })
```
Comment (yogurtandjam) [inline src/commonMain/kotlin/exchange.dydx.abacus/state/model/TradingStateMachine+TransferInput.kt:251]: [fix] remove this

File: src/commonMain/kotlin/exchange.dydx.abacus/utils/String+Utils.kt [hunk 5]
```diff
@@ -24,6 +24,29 @@ fun String.toNobleAddress(): String? {
     }
 }
 
+fun String.toOsmosisAddress(): String? {
```
Comment (yogurtandjam) [inline src/commonMain/kotlin/exchange.dydx.abacus/utils/String+Utils.kt:27]: ask skip about how fallback addresses work - do we need to add these wallets to our autosweeping logic?

File: src/commonMain/kotlin/exchange.dydx.abacus/state/v2/supervisor/OnboardingSupervisor.kt [hunk 6]
```diff
@@ -1011,12 +1030,17 @@ internal class OnboardingSupervisor(
...
-            "dest_asset_denom" to toToken,
+            "dest_asset_denom" to toTokenDenomForAPIUse,
             "dest_asset_chain_id" to toChain,
             "chain_ids_to_addresses" to mapOf(
-                fromChain to sourceAddress,
-                toChain to accountAddress,
+                fromChain to accountAddress,
+                osmosisChainId to osmosisAddress,
+                nobleChainId to nobleAddress,
+                "neutron-1" to accountAddress.toNeutronAddress(),
```
Comment (yogurtandjam) [inline src/commonMain/kotlin/exchange.dydx.abacus/state/v2/supervisor/OnboardingSupervisor.kt:1039]: fixed in a followup PR

**Review by rosepuppy — APPROVED** [review]

File: src/commonMain/kotlin/exchange.dydx.abacus/utils/String+Utils.kt (hunk 5 above)
Comment (yogurtandjam) [reply src/commonMain/kotlin/exchange.dydx.abacus/utils/String+Utils.kt:27]: confirmed it's at parity. they do not automatically sweep back into noble but not worried about that rn

**Pull request merged**